services/autoscaling/groups.go

```diff
@@ -0,0 +1,544 @@
+	utils.SetPropWithNilCheck(scaleIn, "termination_policy_type", scaleInAction.TerminationPolicy)
+	utils.SetPropWithNilCheck(scaleIn, "cooldown_period", scaleInAction.CooldownPeriod)
+	utils.SetPropWithNilCheck(scaleIn, "delete_volumes", scaleInAction.DeleteVolumes)
+
+	return scaleIn
+}
+
+func setScaleOutActionProperties(scaleOutAction autoscaling.GroupPolicyScaleOutAction) map[string]any {
+
+	scaleOut := map[string]any{}
+
+	utils.SetPropWithNilCheck(scaleOut, "amount", scaleOutAction.Amount)
+	utils.SetPropWithNilCheck(scaleOut, "amount_type", scaleOutAction.AmountType)
+	utils.SetPropWithNilCheck(scaleOut, "cooldown_period", scaleOutAction.CooldownPeriod)
+
+	return scaleOut
+}
+
+func setReplicaConfiguration(d *schema.ResourceData, replicaConfiguration autoscaling.ReplicaPropertiesPost) map[string]any {
+
+	replica := map[string]any{}
+
+	utils.SetPropWithNilCheck(replica, "availability_zone", replicaConfiguration.AvailabilityZone)
+	utils.SetPropWithNilCheck(replica, "cores", replicaConfiguration.Cores)
+	utils.SetPropWithNilCheck(replica, "cpu_family", replicaConfiguration.CpuFamily)
+	utils.SetPropWithNilCheck(replica, "ram", replicaConfiguration.Ram)
+
+	if replicaConfiguration.Nics != nil {
+		var nics []any
+		for _, nic := range *replicaConfiguration.Nics {
+			nicEntry := setNicProperties(nic)
+			nics = append(nics, nicEntry)
+		}
+		replica["nic"] = nics
+	}
+
+	if replicaConfiguration.Volumes != nil {
+		var volumes []any
+		for idx, volume := range *replicaConfiguration.Volumes {
+			volumeEntry := setVolumeProperties(d, idx, volume)
+			volumes = append(volumes, volumeEntry)
+		}
+		replica["volume"] = volumes
+	}
+
+	return replica
+}
+
+func setNicProperties(replicaNic autoscaling.ReplicaNic) map[string]any {
+	nic := map[string]any{}
+
+	utils.SetPropWithNilCheck(nic, "lan", replicaNic.Lan)
+	utils.SetPropWithNilCheck(nic, "name", replicaNic.Name)
+	utils.SetPropWithNilCheck(nic, "dhcp", replicaNic.Dhcp)
+
+	return nic
+}
+
+func setVolumeProperties(d *schema.ResourceData, index int, replicaVolume autoscaling.ReplicaVolumePost) map[string]any {
+	volume := map[string]any{}
+
+	utils.SetPropWithNilCheck(volume, "image", replicaVolume.Image)
+	utils.SetPropWithNilCheck(volume, "image_alias", replicaVolume.ImageAlias)
+	utils.SetPropWithNilCheck(volume, "name", replicaVolume.Name)
+	utils.SetPropWithNilCheck(volume, "size", replicaVolume.Size)
+	utils.SetPropWithNilCheck(volume, "type", replicaVolume.Type)
+	utils.SetPropWithNilCheck(volume, "boot_order", replicaVolume.BootOrder)
+	utils.SetPropWithNilCheck(volume, "bus", replicaVolume.Bus)
+	//we need to take these from schema as they are not returned by API
+	volumeMap := d.Get("replica_configuration.0.volume").(*schema.Set).List()[index].(map[string]any)
+
+	volume["image_password"] = volumeMap["image_password"]
+	volume["ssh_keys"] = volumeMap["ssh_keys"]
+	volume["user_data"] = volumeMap["user_data"]
```

Comment:
This will panic for the data source, since the data source schema will be empty initially
```suggestion
	volumeMap, ok := d.GetOk("replica_configuration.0.volume")
	if ok {
		volumeMap := (volumeMap).(*schema.Set).List()[index].(map[string]any)
		volume["image_password"] = volumeMap["image_password"]
		volume["ssh_keys"] = volumeMap["ssh_keys"]
		volume["user_data"] = volumeMap["user_data"]
	}
```